server/src/server.ts

```diff
@@ -291,22 +296,44 @@ export default class BashServer {
   }
 
   private onCompletion(params: LSP.TextDocumentPositionParams): BashCompletionItem[] {
-    const word = this.getWordAtPoint(params)
+    const word = this.getWordAtPoint({
+      ...params,
+      position: {
+        line: params.position.line,
+        // Go one character back to get completion on the current word
+        character: Math.max(params.position.character - 1, 0),
+      },
+    })
     this.logRequest({ request: 'onCompletion', params, word })
 
+    if (word && word.startsWith('#')) {
+      // Inside a comment block
+      return []
+    }
+
     const currentUri = params.textDocument.uri
 
+    const shouldCompleteOnVariables = word
+      ? PARAMETER_EXPANSION_PREFIXES.has(word)
+      : false
+
     const symbolCompletions =
       word === null
         ? []
         : this.getCompletionItemsForSymbols({
-            symbols: this.analyzer.findSymbolsMatchingWord({
-              exactMatch: false,
-              word,
-            }),
+            symbols: shouldCompleteOnVariables
+              ? this.analyzer.getAllVariableSymbols()
+              : this.analyzer.findSymbolsMatchingWord({
+                  exactMatch: false,
+                  word,
+                }),
             currentUri,
           })
 
+    if (shouldCompleteOnVariables) {
```

Comment:
is this an optimization to avoid creating the massive `allCompletions` array below?
If so, this if statement only returns when the user type:

```
echo $
echo ${
echo "$
echo "${
```

and if the user types anything after that, the value of `shouldCompleteOnVariables` will be false.
Is this desired behavior, or do we want `shouldCompleteOnVariables` to be true when the user has something like:

```sh
start="some variable"
# user begins to type 'st':
# and hopes to autocomplete 'start'
echo ${st

# shouldCompleteOnVariables is false though
```

Reply:
Thanks for reading through this carefully.

> is this an optimization to avoid creating the massive allCompletions array below?

This is actually not an optimization. The reason for the early exit is that we only want to suggest parameters/variables here if the word that we complete on is `$` or `${`.

> and if the user types anything after that, the value of shouldCompleteOnVariables will be false.

From my experience, this is not how language server protocol clients work. They get completion for a word (e.g. `$`) once, meaning they do not continuously call the onCompletion while the user types. This makes our life a bit easier here, as we do not need to adjust while the user types out a word.